File: console/python/test/verbs.py
import json
import os
import subprocess
import tempfile
import unittest

import gmr


def a_repository():
    root = tempfile.mkdtemp()
    subprocess.run(["git", "init", "-q", root], check=True)
    os.makedirs(os.path.join(root, "envs"), exist_ok=True)
    os.makedirs(os.path.join(root, "memories"), exist_ok=True)
    with open(os.path.join(root, "envs", "prod.yaml"), "w") as f:
        f.write("service:\n  replicas: 9\n")
    with open(os.path.join(root, "memories", "replicas.md"), "w") as f:
        f.write("Nine, because eight cannot survive a rolling restart.\n")
    return root


def opened(root):
    return gmr.open(
        {
            "root": root,
            "providers": {"git": True},
            "recipes": {
                "file": {
                    "replicas": {"path": "envs/{env}.yaml", "select": "$.service.replicas"}
                }
            },
        }
    )


def replicated(g):
    g.open(
        {
            "key": "prod-replicas",
            "probe": {"kind": "file", "name": "replicas"},
            "initial": {"position": {"env": "prod"}},
            "transitions": [
                {"when": "true", "to": "{ position: state.position, v: obs.value }"}
            ],
        }
    )


class Verbs(unittest.TestCase):
    def test_the_module_names_the_contract_it_serves(self):
        self.assertEqual(gmr.CONTRACT, "gmr.contract.v10")

    def test_five_lines_get_a_sentences_grounding(self):
        g = opened(a_repository())
        replicated(g)
        g.bind("git:memories/replicas.md", ["prod-replicas"], "derived")

        standing = g.ground(["git:memories/replicas.md"], {"max_staleness_ms": 0})[0]
        self.assertEqual(standing["claim"]["provider"], "git")
        self.assertEqual(len(standing["on"]), 1, "the sentence is about one anchor")
        self.assertEqual(standing["on"][0]["anchored"], "on")
        self.assertIn(
            standing["on"][0]["warrant"]["holding"]["holding"],
            ["holds", "moved", "incomparable", "absent", "never_established", "undated"],
        )

    def test_an_address_that_names_no_store_is_refused_before_anything_is_asked(self):
        g = opened(a_repository())
        with self.assertRaises(gmr.Fault) as refused:
            g.ground(["memories/replicas.md"])
        spoken = str(refused.exception)
        self.assertTrue(
            spoken.startswith("refused: "),
            f"the kind is a token in front of the prose: {spoken}",
        )
        self.assertIn("names nothing", spoken)

    def test_read_hands_back_the_envelope_and_a_carried_edge_says_who_asserted_it(self):
        g = opened(a_repository())
        replicated(g)
        g.bind("git:memories/replicas.md", ["prod-replicas"], "derived")
        g.link("git:memories/replicas.md", "git:memories/why.md", "rests-on", "adjudicated")

        view = g.read("prod-replicas")
        self.assertEqual(view["key"], "prod-replicas")
        self.assertEqual(len(view["memories"]), 1, "carry is opt-in")
        held = view["memories"][0]
        self.assertTrue(held["grounded"])
        self.assertEqual(held["links"], [
            {"to": {"provider": "git", "external_id": "memories/why.md"},
             "kind": "rests-on", "source": "adjudicated"}
        ])

        dropped = g.unlink(
            from_="git:memories/replicas.md",
            to="git:memories/why.md",
            kind="rests-on",
            source="adjudicated",
        )
        self.assertEqual(dropped, 1)
        after = g.read("prod-replicas")
        self.assertEqual(after["memories"][0]["links"], [])

    def test_what_changed_since_a_cursor_comes_back_as_edges(self):
        g = opened(a_repository())
        replicated(g)
        seen = g.since(0)
        self.assertIsInstance(seen["edges"], list)


if __name__ == "__main__":
    unittest.main(verbosity=1)

File: console/node/test/verbs.mjs
import { createRequire } from "node:module";
import { execFileSync } from "node:child_process";
import { mkdtempSync, mkdirSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import test from "node:test";
import assert from "node:assert/strict";

const require = createRequire(import.meta.url);
const native = require(process.env.GMR_ADDON ?? "../gmr.node");

function aRepository() {
  const root = mkdtempSync(join(tmpdir(), "gmr-node-"));
  execFileSync("git", ["init", "-q"], { cwd: root });
  mkdirSync(join(root, "envs"), { recursive: true });
  mkdirSync(join(root, "memories"), { recursive: true });
  writeFileSync(join(root, "envs", "prod.yaml"), "service:\n  replicas: 9\n");
  writeFileSync(
    join(root, "memories", "replicas.md"),
    "Nine, because eight cannot survive a rolling restart.\n",
  );
  return root;
}

async function opened(root) {
  return native.open({
    root,
    providers: { git: true },
    recipes: {
      file: { replicas: { path: "envs/{env}.yaml", select: "$.service.replicas" } },
    },
  });
}

test("five lines get a sentence's grounding", async () => {
  const root = aRepository();
  const gmr = await opened(root);

  await gmr.open({
    key: "prod-replicas",
    probe: { kind: "file", name: "replicas" },
    initial: { position: { env: "prod" } },
    transitions: [
      { when: "true", to: "{ position: state.position, v: obs.value }" },
    ],
  });

  await gmr.bind("git:memories/replicas.md", ["prod-replicas"], "derived");

  const [standing] = await gmr.ground(["git:memories/replicas.md"], {
    max_staleness_ms: 0,
  });

  assert.equal(standing.claim.provider, "git");
  assert.equal(standing.on.length, 1, "the sentence is about one anchor");
  assert.equal(standing.on[0].anchored, "on");
  assert.equal(standing.on[0].key, "prod-replicas");
  assert.ok(standing.on[0].warrant, "and the anchor says whether it still holds");
  assert.ok(
    standing.on[0].evidence.reading,
    "with an address for what was read, and no value: a caller that wants the fact " +
      "asks the fact layer for it, and one that wants to audit the judgement follows " +
      "the address",
  );
  assert.equal(standing.on[0].evidence.value, undefined);

  const declared = ["current", "unverified", "rewritten", "gone", "no_provider", "unreachable"];
  assert.ok(
    declared.includes(standing.record.grounding),
    "index.d.ts is hand-written, because the alternative is declaring every " +
      "contract type a second time in Rust. So the discriminants it names have " +
      "to be walked by something: " + JSON.stringify(standing.record),
  );
  assert.ok(
    ["holds", "moved", "incomparable", "absent", "never_established", "undated"]
      .includes(standing.on[0].warrant.holding.holding),
    JSON.stringify(standing.on[0].warrant),
  );
  assert.ok(
    ["seen", "blind"].includes(standing.on[0].warrant.knowledge.knowledge),
    JSON.stringify(standing.on[0].warrant),
  );
});

test("what changed since a cursor comes back as edges", async () => {
  const root = aRepository();
  const gmr = await opened(root);
  await gmr.open({
    key: "prod-replicas",
    probe: { kind: "file", name: "replicas" },
    initial: { position: { env: "prod" } },
  });

  const seen = await gmr.since(0);
  assert.ok(Array.isArray(seen.edges));
  assert.ok(seen.cursor >= 0);
});

test("an instruction nobody here understands is refused, not dropped", async () => {
  const root = aRepository();
  const gmr = await opened(root);
  await assert.rejects(
    () => gmr.ground(["git:memories/replicas.md"], { maxStaleness: 60000 }),
    /maxStaleness/,
    "a freshness bound silently dropped is an answer served stale under a bound the " +
      "caller believes they set",
  );
});

test("an address that names no store is refused before anything is asked", async () => {
  const root = aRepository();
  const gmr = await opened(root);
  await assert.rejects(
    () => gmr.ground(["memories/replicas.md"]),
    /^.*refused: .*names nothing/,
    "the kind travels in front of the prose, so a caller matches a token, not a sentence",
  );
});

test("read hands back the envelope, and a carried edge says who asserted it", async () => {
  const root = aRepository();
  const gmr = await opened(root);
  await gmr.open({
    key: "prod-replicas",
    probe: { kind: "file", name: "replicas" },
    initial: { position: { env: "prod" } },
    transitions: [
      { when: "true", to: "{ position: state.position, v: obs.value }" },
    ],
  });
  await gmr.bind("git:memories/replicas.md", ["prod-replicas"], "derived");
  await gmr.link("git:memories/replicas.md", "git:memories/why.md", "rests-on", "adjudicated");

  const view = await gmr.read("prod-replicas");
  assert.equal(view.key, "prod-replicas");
  assert.equal(view.memories.length, 1, "carry is opt-in; only the bound record answers");
  const held = view.memories[0];
  assert.equal(held.grounded, true);
  assert.equal(held.links.length, 1);
  assert.equal(held.links[0].kind, "rests-on");
  assert.equal(held.links[0].source, "adjudicated");
  assert.ok(held.warrant, "a bound record carries how it stands to this anchor");

  const dropped = await gmr.unlink("git:memories/replicas.md", "git:memories/why.md", "rests-on", "adjudicated");
  assert.equal(dropped, 1);
  const after = await gmr.read("prod-replicas");
  assert.equal(after.memories[0].links.length, 0, "a revoked edge stops being served");
});
